tests/run/opaque-immutable-array-xm.scala

```diff
@@ -0,0 +1,51 @@
+import scala.reflect.ClassTag
+object Test extends App {
+
+  import java.util.Arrays
+
+  opaque type IArray[A1] = Array[A1]
+
+  implicit object IArray {
+    inline def initialize[A](body: => Array[A]): IArray[A] = body
+    def apply[A: ClassTag](xs: A*): IArray[A] = initialize(Array(xs: _*))
+
+    // These should be inline but that does not work currently. Try again
+    // once inliner is moved to ReifyQuotes
+    def (ia: IArray[A]) length[A]: Int = (ia: Array[A]).length
+    def (ia: IArray[A]) apply[A] (i: Int): A = (ia: Array[A])(i)
+
+    // return a sorted copy of the array
+    def sorted[A <: AnyRef : math.Ordering](ia: IArray[A]): IArray[A] = {
+      val arr = Arrays.copyOf(ia, (ia: Array[A]).length)
+      scala.util.Sorting.quickSort(arr)
+      arr
+    }
+
+    // use a standard java method to search a sorted IArray.
+    // (note that this doesn't mutate the array).
+    def binarySearch(ia: IArray[Long], elem: Long): Int =
+      Arrays.binarySearch(ia, elem)
+  }
+
+  // same as IArray.binarySearch but implemented by-hand.
+  //
+  // given a sorted IArray, returns index of `elem`,
+  // or a negative value if not found.
+  def binaryIndexOf(ia: IArray[Long], elem: Long): Int = {
+    var lower: Int = 0
+    var upper: Int = ia.length - 1
+    while (lower <= upper) {
+      val middle = (lower + upper) >>> 1
+      val n = ia(middle)
+
+      if (n == elem) return middle
+      else if (n < elem) lower = middle + 1
+      else upper = middle - 1
+    }
+    -lower - 1
+  }
+
+  val xs: IArray[Long] = IArray(1L, 2L, 3L)
```

Comment:
You might want to exercise `sorted` Here rather than building a sorted array by hand.